Feature/implement authentication UI screens

AhmadKharfan opened the pull request:
Add Welcome & Login Flows

- WelcomeScreen with WelcomeViewModel (uses Unit state) and effects NavigateToLogin/ContinueAsGuest/ExitApp.

- LoginScreen with LoginViewModel (LoginUiState) and effects for navigation, errors & success snackbars.

- UI split into tiny private composables; navigation via NavController + LaunchedEffect.


https://github.com/user-attachments/assets/2bd3615e-311f-4b85-85d3-0c15ab8ec8cb

AhmadKharfan committed:
integrate authentication feature

AhmadKharfan committed:
add login dependency injection module

AhmadKharfan committed:
make button width dynamic in design system

AhmadKharfan committed:
improve TextField state handling and cursor behavior

AhmadKharfan committed:
add new libraries in auth module

AhmadKharfan committed:
implement navigation for authentication feature

AhmadKharfan committed:
implement login screen viewmodel and actions and events

AhmadKharfan committed:
add login screen structure and components

AhmadKharfan committed:
add welcome screen viewmodel and effects

AhmadKharfan committed:
implement welcome screen and its components

AhmadKharfan committed:
add authentication resources

AhmadKharfan committed:
add arabic translations for onboarding and login

AhmadKharfan committed:
Merge branch 'development' into feature/implement-authentication-ui-screens

ElhamFadel commented on design_system/src/main/java/com/sanaa/designsystem/design_system/component/text_field/PasswordTextField.kt:
```diff
@@ -14,125 +13,95 @@ import androidx.compose.foundation.layout.size
+                        modifier = Modifier
+                            .padding(end = 8.dp)
+                            .size(24.dp),
+                        tint = hintColor
```
we need to make composable more responsible and small so try to add that

```
 icon?.let {
        LeadingIcon(painter = it, description = iconDescription, tint = hintColor)
 }
                
           
```     
then 
```

@Composable
private fun LeadingIcon(painter: Painter, description: String?, tint: Color) {
    Icon(
        painter = painter,
        contentDescription = description,
        modifier = Modifier
            .padding(end = 8.dp)
            .size(24.dp),
        tint = tint
    )
}
```
`

haideraljawasim commented on feature/authentication/presentation/build.gradle.kts:
```diff
@@ -15,4 +15,26 @@ dependencies {
+    implementation(projects.designSystem)
+
+    implementation(libs.kotlinx.serialization.json)
+    implementation(libs.material3)
```
why would you add material 3 for feature gradle  file

AhmadKharfan replied:
I know But we add it in all features gradle  now
we don't have any text component in design system

haideraljawasim replied:
You don't need you can use BasicText from compose foundation

AhmadKharfan replied:
No I prefer ty create out design system text component instead of using basic text

haideraljawasim commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/LoginScreen.kt:
```diff
@@ -0,0 +1,131 @@
+    }
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun LoginContent(
+    state: LoginUiState,
```
please remove anything related to material 3 from the feature you can achieve  anything material 3 provide for you using composables

AhmadKharfan replied:
I’m not using any Material 3 components here just our text composable like in the other features. We should remove all material3 imports from each feature module and create it in the design‑system module instead

haideraljawasim replied:
We already did that in refactor for search and doing it now for details

AhmadKharfan replied:
Great do it also in auth 😉

haideraljawasim commented on design_system/src/main/java/com/sanaa/designsystem/design_system/component/text_field/NovixBasicTextField.kt:
```diff
@@ -32,9 +37,25 @@ internal fun NovixBasicTextField(
     interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
     content: @Composable (innerTextField: @Composable () -> Unit, hintColor: Color) -> Unit,
 ) {
+    var internalValue by rememberSaveable(stateSaver = TextFieldValue.Saver) {
+        mutableStateOf(TextFieldValue(text = value, selection = TextRange(value.length)))
+    }
+
+    LaunchedEffect(value) {
+        if (value != internalValue.text) {
+            internalValue = internalValue.copy(
+                text = value,
+                selection = TextRange(value.length)
+            )
+        }
+    }
+
     NovixBasicTextField(
```
I don't get it why you did all these changes on the text field from design system that already used and not have problem with it

AhmadKharfan replied:
Could you try running the previous version on your emulator? You’ll see that the text cursor stays at the very start of the field, never moves to the end of what you type, and you can’t delete any characters you’ve entered

haideraljawasim commented on design_system/src/main/java/com/sanaa/designsystem/design_system/component/text_field/PasswordTextField.kt:
```diff
@@ -14,125 +13,95 @@ import androidx.compose.foundation.layout.size
-            )
-            PasswordTextField(
-                value = text,
-                onValueChange = { text = it },
```
same here would would you need to delete the Preview !!

haideraljawasim commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/LoginScreen.kt:
```diff
@@ -0,0 +1,131 @@
+fun LoginScreen(
+    modifier: Modifier = Modifier
+) {
+    val viewModel: LoginViewModel = koinViewModel()
```
I prefer injecting the view model in contractor , as it decouples the ui from di  , and more easy to test

AhmadKharfan committed:
Merge branch 'development' into feature/implement-authentication-ui-screens

AhmadKharfan committed:
extract leading icon to a private function
and type preview

AhmadKharfan committed:
inject view models via parameters in login and welcome screen and remove material 3 annotation

This commit refactors the LoginScreen and WelcomeScreen composables to receive their respective ViewModels as parameters, rather than instantiating them directly within the composable function.

This change improves testability and follows recommended practices for dependency injection in Jetpack Compose.

Additionally, the `@OptIn(ExperimentalMaterial3Api::class)` annotation has been removed from `LoginContent` and `WelcomeContent` as it was no longer necessary.

raouf-kamel reviewed (commented):
Please add keyboardActions with onNext to the username text field for better UX

AhmadKharfan committed:
ignore login and welcome screens in codecov

AhmadKharfan committed:
Merge branch 'development' into feature/implement-authentication-ui-screens

ZainabAlnajjar commented on feature/authentication/presentation/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,68 @@
+    <string name="genre_western">Western</string>
+    <string name="genre_war_and_politics">War&amp;Politics</string>
+    <string name="genre_sci_fi_and_fantasy">Sci-Fi&amp;Fantasy</string>
+    <string name="genre_action_and_adventure">Action&amp;Adventure</string>
```
we already refactored the app to support getting genres from api, please remove these strings and fetch the genres from the api if you need it

ZainabAlnajjar replied:
I think you copied another feature strings, please remove all the strings you don't need in your feature

ZainabAlnajjar commented on feature/authentication/presentation/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,68 @@
+    <string name="app_icon">app icon</string>
+    <string name="onboarding_background">onboarding background</string>
+    <string name="welcome_to_novix">Welcome to Novix</string>
+    <string name="discover_the_world_s_greatest_movies_and_series_search_watch_and_save_your_favourites_all_in_one_place">Discover the world’s greatest movies and series. Search, watch, and save your favourites all in one place.</string>
```
this is wrong naming, you can name it `onboarding_caption `or something related to what it is

ZainabAlnajjar commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/welcome/components/WelcomeHero.kt:
```diff
@@ -0,0 +1,56 @@
+import com.sanaa.feature.authentication.presentation.R
+
+@Composable
+fun WelcomeHero(modifier: Modifier = Modifier) {
```
the name is not suitable for composable, I suggest renaming it to `WelcomeSection` or something clearer for developer

AhmadKharfan committed:
Merge branch 'development' into feature/implement-authentication-ui-screens

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/LoginScreen.kt:
```diff
@@ -0,0 +1,129 @@
+        Column(modifier = modifier) {
+            LoginTopBar(onBackClick = listener::onBackClicked)
+
+            Spacer(Modifier.height(12.dp))
```
I think you should use padding top instead of the spacer

AimanYosofiUlt commented on design_system/src/main/java/com/sanaa/designsystem/design_system/component/button/common/ButtonContainer.kt:
```diff
@@ -38,30 +41,24 @@ internal fun ButtonContainer(
     ) {
         content()
 
+        Spacer(modifier = Modifier.width(8.dp))
```
> I think you should use padding top instead of the spacer

Here too

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/components/LoginActions.kt:
```diff
@@ -0,0 +1,49 @@
+                .align(Alignment.CenterHorizontally),
+            isLoading = isLoading
+        )
+        Spacer(Modifier.height(12.dp))
```
> I think you should use padding top instead of the spacer

Here too

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/components/LoginHeader.kt:
```diff
@@ -0,0 +1,37 @@
+            contentDescription = stringResource(R.string.app_icon),
+            modifier = Modifier.size(64.dp)
+        )
+        Spacer(Modifier.height(16.dp))
```
> I think you should use padding top instead of the spacer

Here too

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/components/SignupFooter.kt:
```diff
@@ -0,0 +1,41 @@
+            style = Theme.textStyle.body.small,
+            color = Theme.colors.body
+        )
+        Spacer(Modifier.width(4.dp))
```
> I think you should use padding top instead of the spacer

Here too

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/welcome/components/WelcomeFooter.kt:
```diff
@@ -0,0 +1,74 @@
+        horizontalAlignment = Alignment.CenterHorizontally
+    ) {
+        WelcomeInfo()
+        Spacer(Modifier.height(32.dp))
```
> I think you should use padding top instead of the spacer

Here too

AimanYosofiUlt commented on app/src/main/java/com/sanaa/novix/di/login_modules/loginModule.kt:
```diff
@@ -0,0 +1,9 @@
+
+import org.koin.dsl.module
+
+val loginModule = module {
```
This would be an authentication module instead of loginModule

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/components/LoginActions.kt:
```diff
@@ -0,0 +1,49 @@
+import com.sanaa.feature.authentication.presentation.R
+
+@Composable
+fun LoginActions(
```
We need a Preview for this

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/components/LoginTopBar.kt:
```diff
@@ -0,0 +1,28 @@
+import com.sanaa.feature.authentication.presentation.R
+
+@Composable
+fun LoginTopBar(onBackClick: () -> Unit) {
```
> We need a Preview for this

Here too

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/components/PasswordField.kt:
```diff
@@ -0,0 +1,31 @@
+import com.sanaa.feature.authentication.presentation.R
+
+@Composable
+fun PasswordField(
```
> We need a Preview for this

Here too

ElhamFadel replied:
we will make it at the morning after connect everthing

AimanYosofiUlt commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/welcome/WelcomeScreen.kt:
```diff
@@ -0,0 +1,84 @@
+}
+
+@Composable
+fun WelcomeContent(
```
> We need a Preview for this

Here too

ElhamFadel committed:
remove strings not related

ElhamFadel committed:
rename to welcome section

ZainabAlnajjar commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/login/LoginScreen.kt:
```diff
@@ -0,0 +1,129 @@
+    NovixScaffold(
+        backgroundShapes = { NovixBackgroundShapes() }
+    ) {
+        Column(modifier = modifier) {
+            LoginTopBar(onBackClick = listener::onBackClicked)
```
you are missing 12.dp top padding between the TopBar and the edge of the statusBar (not innerPadding)

ElhamFadel committed:
used padding instead of spacer

ElhamFadel committed:
rename onboarding_background to onboarding_caption

ElhamFadel committed:
added statusbar padding

ZainabAlnajjar commented on feature/authentication/presentation/src/main/java/com/sanaa/presentation/screen/welcome/components/WelcomeHero.kt:
please also rename the file name to `WelcomeSection`

ElhamFadel committed:
changed name of fileWelcomeHero to WelcomeSection and replace spacer to padding

ElhamFadel committed:
Merge branch 'feature/auth/prepare-domain-and-repository' into feature/implement-authentication-ui-screens

# Conflicts:
#	app/src/main/java/com/sanaa/novix/di/appModule.kt
#	feature/authentication/presentation/build.gradle.kts
#	feature/authentication/presentation/src/main/java/com/sanaa/presentation/navigation/AuthNavHost.kt
#	feature/authentication/presentation/src/main/java/com/sanaa/presentation/navigation/AuthNavigation.kt
#	feature/authentication/presentation/src/main/res/values-ar/strings.xml
#	feature/authentication/presentation/src/main/res/values/strings.xml

ElhamFadel committed:
merge property-domain into this branch
